File: backend/src/main/java/com/hostelhub/controller/BookingController.java
package com.hostelhub.controller;

import com.hostelhub.config.UserDetailsImpl;
import com.hostelhub.dto.BookingRequest;
import com.hostelhub.dto.BookingStatusRequest;
import com.hostelhub.entity.*;
import com.hostelhub.repository.BookingRepository;
import com.hostelhub.repository.RoomRepository;
import com.hostelhub.repository.UserRepository;
import com.hostelhub.repository.InvoiceRepository;
import java.time.LocalDate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.HashSet;
import java.util.stream.Collectors;

@CrossOrigin(origins = "*", maxAge = 3600)
@RestController
@RequestMapping("/api/bookings")
public class BookingController {

    @Autowired
    private BookingRepository bookingRepository;

    @Autowired
    private RoomRepository roomRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private InvoiceRepository invoiceRepository;

    @PostMapping
    @PreAuthorize("hasRole('STUDENT')")
    public ResponseEntity<?> createBooking(@RequestBody BookingRequest bookingRequest) {
        UserDetailsImpl userPrincipal = (UserDetailsImpl) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        User student = userRepository.findById(userPrincipal.getId()).orElse(null);

        if (student == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Student not found");
        }

        Optional<Room> roomData = roomRepository.findById(bookingRequest.getRoomId());
        if (roomData.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Room not found");
        }

        Room room = roomData.get();
        if (room.getStatus() != RoomStatus.AVAILABLE) {
            return ResponseEntity.badRequest().body("Room is not available for booking");
        }

        Booking booking = new Booking(student, 
                                      room, 
                                      bookingRequest.getStartDate(), 
                                      bookingRequest.getEndDate());

        Booking savedBooking = bookingRepository.save(booking);
        return ResponseEntity.status(HttpStatus.CREATED).body(savedBooking);
    }

    @GetMapping("/student")
    @PreAuthorize("hasRole('STUDENT')")
    public ResponseEntity<List<Booking>> getStudentBookings() {
        UserDetailsImpl userPrincipal = (UserDetailsImpl) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        List<Booking> bookings = bookingRepository.findByStudentId(userPrincipal.getId());
        return ResponseEntity.ok(bookings);
    }

    @GetMapping("/owner")
    @PreAuthorize("hasRole('OWNER')")
    public ResponseEntity<List<Booking>> getOwnerBookings() {
        UserDetailsImpl userPrincipal = (UserDetailsImpl) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        List<Booking> bookings = bookingRepository.findByOwnerId(userPrincipal.getId());
        return ResponseEntity.ok(bookings);
    }

    @PutMapping("/{id}/status")
    @PreAuthorize("hasRole('OWNER')")
    public ResponseEntity<?> updateBookingStatus(@PathVariable Long id, @RequestBody BookingStatusRequest statusRequest) {
        UserDetailsImpl userPrincipal = (UserDetailsImpl) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        Optional<Booking> bookingData = bookingRepository.findById(id);

        if (bookingData.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Booking not found");
        }

        Booking booking = bookingData.get();
        // Check ownership of the hostel containing the room
        if (!booking.getRoom().getHostel().getOwner().getId().equals(userPrincipal.getId())) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).body("You do not own this hostel listing");
        }

        booking.setStatus(statusRequest.getStatus());

        // Business logic: if accepted, we can mark the room occupied and generate monthly invoices
        if (statusRequest.getStatus() == BookingStatus.ACCEPTED) {
            Room room = booking.getRoom();
            room.setStatus(RoomStatus.OCCUPIED);
            roomRepository.save(room);

            // 1. Generate Security Deposit (Advance) Invoice (1 Month Rent)
            Invoice securityDeposit = new Invoice(
                booking,
                booking.getRoom().getPricePerMonth(),
                booking.getStartDate(),
                InvoiceStatus.PENDING,
                "Security Deposit (Advance)"
            );
            invoiceRepository.save(securityDeposit);

            // 2. Pre-generate monthly invoices for the booking duration
            LocalDate current = booking.getStartDate();
            LocalDate end = booking.getEndDate();
            while (current.isBefore(end) || (current.getMonthValue() == end.getMonthValue() && current.getYear() == end.getYear())) {
                LocalDate dueDate = current.withDayOfMonth(1);
                if (dueDate.isBefore(booking.getStartDate())) {
                    dueDate = booking.getStartDate();
                }
                
                String billingMonth = current.getMonth().name() + " " + current.getYear();
                Invoice invoice = new Invoice(booking, 
                                              booking.getRoom().getPricePerMonth(), 
                                              dueDate, 
                                              InvoiceStatus.PENDING, 
                                              billingMonth);
                invoiceRepository.save(invoice);
                
                // Advance to next month
                current = current.plusMonths(1).withDayOfMonth(1);
            }
        } else if (statusRequest.getStatus() == BookingStatus.CANCELLED || statusRequest.getStatus() == BookingStatus.REJECTED) {
            Room room = booking.getRoom();
            // Optional: reset room status if needed, but only if it's currently occupied
            if (room.getStatus() == RoomStatus.OCCUPIED) {
                room.setStatus(RoomStatus.AVAILABLE);
                roomRepository.save(room);
            }
        }

        Booking updatedBooking = bookingRepository.save(booking);
        return ResponseEntity.ok(updatedBooking);
    }

    @PutMapping("/{id}/terminate")
    @PreAuthorize("hasRole('OWNER')")
    public ResponseEntity<?> terminateBooking(@PathVariable Long id) {
        UserDetailsImpl userPrincipal = (UserDetailsImpl) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        Optional<Booking> bookingData = bookingRepository.findById(id);

        if (bookingData.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Booking not found");
        }

        Booking booking = bookingData.get();
        if (!booking.getRoom().getHostel().getOwner().getId().equals(userPrincipal.getId())) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).body("You do not own this hostel listing");
        }

        booking.setStatus(BookingStatus.CANCELLED);
        bookingRepository.save(booking);

        Room room = booking.getRoom();
        room.setStatus(RoomStatus.AVAILABLE);
        roomRepository.save(room);

        // Delete outstanding pending invoices
        List<Invoice> invoices = invoiceRepository.findAll().stream()
            .filter(inv -> inv.getBooking().getId().equals(booking.getId()))
            .collect(Collectors.toList());
        for (Invoice inv : invoices) {
            if (inv.getStatus() == InvoiceStatus.PENDING) {
                invoiceRepository.delete(inv);
            }
        }

        return ResponseEntity.ok().body("Booking terminated and room released successfully.");
    }

    @PutMapping("/{id}/checkout-date")
    @PreAuthorize("hasRole('STUDENT')")
    public ResponseEntity<?> changeCheckoutDate(@PathVariable Long id, @RequestBody java.util.Map<String, String> payload) {
        UserDetailsImpl userPrincipal = (UserDetailsImpl) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        Optional<Booking> bookingData = bookingRepository.findById(id);

        if (bookingData.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Booking not found");
        }

        Booking booking = bookingData.get();
        if (!booking.getStudent().getId().equals(userPrincipal.getId())) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).body("You do not own this booking");
        }

        String endDateStr = payload.get("endDate");
        if (endDateStr == null) {
            return ResponseEntity.badRequest().body("endDate is required");
        }

        LocalDate newEndDate = LocalDate.parse(endDateStr);
        if (newEndDate.isBefore(booking.getStartDate())) {
            return ResponseEntity.badRequest().body("Checkout date cannot be before check-in date");
        }

        booking.setEndDate(newEndDate);
        bookingRepository.save(booking);

        // Recalculate Invoices
        List<Invoice> existing = invoiceRepository.findAll().stream()
            .filter(inv -> inv.getBooking().getId().equals(booking.getId()))
            .collect(Collectors.toList());

        Set<String> paidMonths = existing.stream()
            .filter(inv -> inv.getStatus() == InvoiceStatus.PAID)
            .map(Invoice::getBillingMonth)
            .collect(Collectors.toSet());

        // Delete pending monthly invoices
        for (Invoice inv : existing) {
            if (inv.getStatus() == InvoiceStatus.PENDING && !inv.getBillingMonth().equals("Security Deposit (Advance)")) {
                invoiceRepository.delete(inv);
            }
        }

        // Re-generate monthly invoices up to the new endDate
        LocalDate current = booking.getStartDate();
        LocalDate end = booking.getEndDate();
        while (current.isBefore(end) || (current.getMonthValue() == end.getMonthValue() && current.getYear() == end.getYear())) {
            String billingMonth = current.getMonth().name() + " " + current.getYear();
            if (!paidMonths.contains(billingMonth)) {
                LocalDate dueDate = current.withDayOfMonth(1);
                if (dueDate.isBefore(booking.getStartDate())) {
                    dueDate = booking.getStartDate();
                }
                Invoice invoice = new Invoice(booking, 
                                              booking.getRoom().getPricePerMonth(), 
                                              dueDate, 
                                              InvoiceStatus.PENDING, 
                                              billingMonth);
                invoiceRepository.save(invoice);
            }
            current = current.plusMonths(1).withDayOfMonth(1);
        }

        return ResponseEntity.ok(booking);
    }
}

File: backend/src/controllers/booking.controller.js
import prisma from '../db.js';

export const createBooking = async (req, res) => {
  const { roomId, startDate, endDate } = req.body;
  const studentId = req.userId;

  try {
    const student = await prisma.user.findUnique({
      where: { id: studentId }
    });

    if (!student) {
      return res.status(401).json({ message: "Student not found" });
    }

    const room = await prisma.room.findUnique({
      where: { id: parseInt(roomId) }
    });

    if (!room) {
      return res.status(404).json({ message: "Room not found" });
    }

    if (room.status !== 'AVAILABLE') {
      return res.status(400).json({ message: "Room is not available for booking" });
    }

    const booking = await prisma.booking.create({
      data: {
        studentId,
        roomId: parseInt(roomId),
        startDate: new Date(startDate),
        endDate: new Date(endDate),
        status: 'PENDING'
      }
    });

    return res.status(201).json(booking);
  } catch (error) {
    console.error("Create Booking Error:", error);
    return res.status(500).json({ message: "Internal server error." });
  }
};

export const getStudentBookings = async (req, res) => {
  const studentId = req.userId;

  try {
    const bookings = await prisma.booking.findMany({
      where: { studentId },
      include: {
        room: {
          include: {
            hostel: true
          }
        }
      }
    });
    return res.status(200).json(bookings);
  } catch (error) {
    console.error("Get Student Bookings Error:", error);
    return res.status(500).json({ message: "Internal server error." });
  }
};

export const getOwnerBookings = async (req, res) => {
  const ownerId = req.userId;

  try {
    const bookings = await prisma.booking.findMany({
      where: {
        room: {
          hostel: {
            ownerId
          }
        }
      },
      include: {
        student: {
          select: {
            id: true,
            username: true,
            email: true,
            name: true,
            phone: true
          }
        },
        room: {
          include: {
            hostel: true
          }
        }
      }
    });
    return res.status(200).json(bookings);
  } catch (error) {
    console.error("Get Owner Bookings Error:", error);
    return res.status(500).json({ message: "Internal server error." });
  }
};

export const updateBookingStatus = async (req, res) => {
  const id = parseInt(req.params.id);
  const { status } = req.body;
  const ownerId = req.userId;

  try {
    const booking = await prisma.booking.findUnique({
      where: { id },
      include: {
        room: {
          include: {
            hostel: true
          }
        }
      }
    });

    if (!booking) {
      return res.status(404).json({ message: "Booking not found" });
    }

    if (booking.room.hostel.ownerId !== ownerId) {
      return res.status(403).json({ message: "You do not own this hostel listing" });
    }

    const updatedBooking = await prisma.booking.update({
      where: { id },
      data: { status }
    });

    if (status === 'ACCEPTED') {
      // 1. Mark room status occupied
      await prisma.room.update({
        where: { id: booking.roomId },
        data: { status: 'OCCUPIED' }
      });

      // 2. Generate Security Deposit Invoice (1 Month Rent)
      await prisma.invoice.create({
        data: {
          bookingId: booking.id,
          amount: booking.room.pricePerMonth,
          dueDate: booking.startDate,
          status: 'PENDING',
          billingMonth: "Security Deposit (Advance)"
        }
      });

      // 3. Pre-generate monthly invoices for booking duration
      let current = new Date(booking.startDate);
      const end = new Date(booking.endDate);

      while (current < end || (current.getMonth() === end.getMonth() && current.getFullYear() === end.getFullYear())) {
        let dueDate = new Date(current.getFullYear(), current.getMonth(), 1);
        if (dueDate < booking.startDate) {
          dueDate = booking.startDate;
        }

        const billingMonth = current.toLocaleString('en-US', { month: 'long' }).toUpperCase() + " " + current.getFullYear();

        await prisma.invoice.create({
          data: {
            bookingId: booking.id,
            amount: booking.room.pricePerMonth,
            dueDate,
            status: 'PENDING',
            billingMonth
          }
        });

        // Advance to next month
        current.setMonth(current.getMonth() + 1);
        current.setDate(1);
      }
    } else if (status === 'CANCELLED' || status === 'REJECTED') {
      // Reset room to available if occupied
      if (booking.room.status === 'OCCUPIED') {
        await prisma.room.update({
          where: { id: booking.roomId },
          data: { status: 'AVAILABLE' }
        });
      }
    }

    return res.status(200).json(updatedBooking);
  } catch (error) {
    console.error("Update Booking Status Error:", error);
    return res.status(500).json({ message: "Internal server error." });
  }
};

export const terminateBooking = async (req, res) => {
  const id = parseInt(req.params.id);
  const ownerId = req.userId;

  try {
    const booking = await prisma.booking.findUnique({
      where: { id },
      include: {
        room: {
          include: {
            hostel: true
          }
        }
      }
    });

    if (!booking) {
      return res.status(404).json({ message: "Booking not found" });
    }

    if (booking.room.hostel.ownerId !== ownerId) {
      return res.status(403).json({ message: "You do not own this hostel listing" });
    }

    // Set status cancelled
    await prisma.booking.update({
      where: { id },
      data: { status: 'CANCELLED' }
    });

    // Set room status available
    await prisma.room.update({
      where: { id: booking.roomId },
      data: { status: 'AVAILABLE' }
    });

    // Delete outstanding pending invoices
    await prisma.invoice.deleteMany({
      where: {
        bookingId: id,
        status: 'PENDING'
      }
    });

    return res.status(200).send("Booking terminated and room released successfully.");
  } catch (error) {
    console.error("Terminate Booking Error:", error);
    return res.status(500).json({ message: "Internal server error." });
  }
};

export const changeCheckoutDate = async (req, res) => {
  const id = parseInt(req.params.id);
  const { endDate } = req.body;
  const studentId = req.userId;

  if (!endDate) {
    return res.status(400).json({ message: "endDate is required" });
  }

  try {
    const booking = await prisma.booking.findUnique({
      where: { id },
      include: {
        room: true
      }
    });

    if (!booking) {
      return res.status(404).json({ message: "Booking not found" });
    }

    if (booking.studentId !== studentId) {
      return res.status(403).json({ message: "You do not own this booking" });
    }

    const newEndDate = new Date(endDate);
    if (newEndDate < booking.startDate) {
      return res.status(400).json({ message: "Checkout date cannot be before check-in date" });
    }

    // Update booking end date
    const updatedBooking = await prisma.booking.update({
      where: { id },
      data: { endDate: newEndDate },
      include: {
        room: true
      }
    });

    // Get existing invoices
    const existingInvoices = await prisma.invoice.findMany({
      where: { bookingId: id }
    });

    const paidMonths = new Set(
      existingInvoices
        .filter(inv => inv.status === 'PAID')
        .map(inv => inv.billingMonth)
    );

    // Delete pending monthly invoices (except security deposit)
    await prisma.invoice.deleteMany({
      where: {
        bookingId: id,
        status: 'PENDING',
        NOT: {
          billingMonth: "Security Deposit (Advance)"
        }
      }
    });

    // Re-generate monthly invoices up to newEndDate
    let current = new Date(booking.startDate);
    const end = new Date(newEndDate);

    while (current < end || (current.getMonth() === end.getMonth() && current.getFullYear() === end.getFullYear())) {
      const billingMonth = current.toLocaleString('en-US', { month: 'long' }).toUpperCase() + " " + current.getFullYear();

      if (!paidMonths.has(billingMonth)) {
        let dueDate = new Date(current.getFullYear(), current.getMonth(), 1);
        if (dueDate < booking.startDate) {
          dueDate = booking.startDate;
        }

        await prisma.invoice.create({
          data: {
            bookingId: id,
            amount: booking.room.pricePerMonth,
            dueDate,
            status: 'PENDING',
            billingMonth
          }
        });
      }

      // Advance to next month
      current.setMonth(current.getMonth() + 1);
      current.setDate(1);
    }

    return res.status(200).json(updatedBooking);
  } catch (error) {
    console.error("Change Checkout Date Error:", error);
    return res.status(500).json({ message: "Internal server error." });
  }
};
